### PR TITLE
Bitwise builtin

### LINKED ISSUE #113: Implement bitwise builtin

### PR DESCRIPTION
**Pull request by mmsc2**

### CODE REVIEW

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: We can remove this field and use an constant instead

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]:
```suggestion
	return "bitwise"
```

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: We don't need this

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: TotalNBits is always 251 so we can use a constant instead

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: This should be <= 1. 
Even better, we could set the constant `INPUT_CELLS_PER_INSTANCE = 2` and use it in this if

File: pkg/vm/memory/memory.go
Comment (fmoletta) [inline pkg/vm/memory/memory.go]: Dont do this

File: pkg/runners/builtin_runner/bitwise.go
Comment (mmsc2) [reply pkg/runners/builtin_runner/bitwise.go]: INPUT_CELLS_PER_INSTANCE shouldn't be 1 ?

File: pkg/vm/memory/memory.go
Comment (mmsc2) [reply pkg/vm/memory/memory.go]: this is from the PR of range check i guess, i had to merge that branch into this one as i needed somethings

File: pkg/runners/builtin_runner/bitwise.go
Comment (mmsc2) [reply pkg/runners/builtin_runner/bitwise.go]: or you mean setting in two, because it contemplates the case <= 1 ?

**Review by fmoletta — COMMENTED** [review]
Looks good, suggested some minor changes but it should be good to merge after adressing them and adding some integration tests
We should add some integration tests that use CairoRun to run a program that uses bitwise operation, you can get some cairo programs from the cairo-vm repo
This PR contains changes from the range_check PR, pease remove them so we can merge it, otherwise we have to wait till range_check is finished and merged

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [reply pkg/runners/builtin_runner/bitwise.go]: Yes, the number of  input cells for bitwise is 2 so adding the constant and using it here will make it easier to document

**Comment by fmoletta** [comment]
In order to use this builtin in a cairo program we should create it in the `NewCairoRunner` function (in cairo_runner.go file), there is a comment there showing where to add them, please add it

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: nit: We can omit this field when creating the struct and it will take the zero-value (which is 0,0)

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]: Remove TODO

File: README.md
Comment (fmoletta) [inline README.md]:
```suggestion
- After the operands are fetched it checks that both of them are felts because bitwise can not be performed on relocatable or nil values.
```

File: README.md
Comment (fmoletta) [inline README.md]:
```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations.:
```

File: README.md
Comment (fmoletta) [inline README.md]:
```suggestion
- If the index is less than the number of input cells then the methods just return nil
```

File: README.md
Comment (fmoletta) [inline README.md:2525]: Add the definition of the constants to this code block

File: README.md
Comment (fmoletta) [inline README.md:2479]: Add BitWiseBuiltinRunner type definton

File: README.md
Comment (fmoletta) [inline README.md]:
```suggestion
#### Implementing each builtin runner 
```
This has the same hierarchy as Write your own Vm title

File: README.md
Comment (fmoletta) [inline README.md]:
```suggestion
##### Bitwise builtin
```

File: pkg/runners/builtin_runner/bitwise.go
Comment (fmoletta) [inline pkg/runners/builtin_runner/bitwise.go]:
```suggestion
const BITWISE_CELLS_PER_INSTANCE = 5
const BITWISE_TOTAL_N_BITS = 251
const BIWISE_INPUT_CELLS_PER_INSTANCE = 2
```
As all the builtins are in the same package, constant's names can clash

File: README.md
Comment (gabrielbosio) [inline README.md:2498]:
```suggestion
For the `InitializeSegments` we just add a segment to the memory and store in the base attribute, which is the first address of the segment. 
```

File: README.md
Comment (gabrielbosio) [inline README.md]:
```suggestion
The method `DeducedMemoryCell` fetches the operands from memory and performs the following operations: 
```

File: README.md
Comment (gabrielbosio) [inline README.md]:
```suggestion
- If the index is less than the number of input cells then the method returns nil
```

File: README.md
Comment (gabrielbosio) [inline README.md]:
```suggestion
- If the number of bits of any operand is greater than `TOTAL_N_BITS` the method fails because we are out of the field.
```

File: pkg/lambdaworks/lambdaworks.go
Comment (gabrielbosio) [inline pkg/lambdaworks/lambdaworks.go]: Why is this function needed for this builtin?

Comment (mmsc2) [reply pkg/lambdaworks/lambdaworks.go]: is not, there was a confusion with the ^ operand that i though that it was exponentiation but it was xor. i left it because maybe will be used in the future

File: pkg/runners/builtin_runner/bitwise.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise.go]:
```suggestion
	x_addr := address.SubUint(index)
```

File: pkg/runners/builtin_runner/bitwise.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise.go]: Please, remove this newline

File: pkg/runners/builtin_runner/bitwise.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise.go]: Please, remove this newline

File: pkg/runners/builtin_runner/bitwise.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise.go]:
```suggestion
			return nil, fmt.Errorf("Expected Intenger x to be smaller than 2^%d", BITWISE_TOTAL_N_BITS)
```

File: pkg/runners/builtin_runner/bitwise.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise.go]:
```suggestion
			return nil, errors.New("Expected Intenger y to be smaller than 2^%d", BITWISE_TOTAL_N_BITS)
```

File: pkg/lambdaworks/lambdaworks.go
Comment (gabrielbosio) [reply pkg/lambdaworks/lambdaworks.go]: If this function is needed, please move this and its tests to a new PR

File: pkg/runners/builtin_runner/bitwise_test.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise_test.go]: I think we can remove this insertion because we would be inserting a zero

File: pkg/runners/builtin_runner/bitwise_test.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise_test.go]: I think we can remove this insertion because we would be inserting a zero

File: pkg/runners/builtin_runner/bitwise_test.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise_test.go]: I think we can remove this insertion because we would be inserting a zero

File: pkg/runners/builtin_runner/bitwise_test.go
Comment (gabrielbosio) [inline pkg/runners/builtin_runner/bitwise_test.go]: I think we can remove this insertion because we would be inserting a zero

File: pkg/runners/cairo_runner.go
Comment (gabrielbosio) [inline pkg/runners/cairo_runner.go]: For clarity, we can move the builtin to a new `builtins` folder

File: README.md
Comment (toni-calvin) [reply README.md:2498]: I think it would be nicer something like: 

```suggestion
In the `InitializeSegments` method, we add a new segment to the memory and set the base attribute as the first segment's address. 
```

File: pkg/lambdaworks/lambdaworks.go
Comment (toni-calvin) [reply pkg/lambdaworks/lambdaworks.go]: Done here: #141

File: pkg/runners/cairo_runner.go
Comment (toni-calvin) [reply pkg/runners/cairo_runner.go]: Done!

File: pkg/runners/builtin_runner/bitwise.go
Comment (toni-calvin) [reply pkg/runners/builtin_runner/bitwise.go]: Done!

File: pkg/runners/builtin_runner/bitwise.go
Comment (toni-calvin) [reply pkg/runners/builtin_runner/bitwise.go]: Done!

File: pkg/runners/builtin_runner/bitwise.go
Comment (toni-calvin) [reply pkg/runners/builtin_runner/bitwise.go]: Done!

File: pkg/runners/builtin_runner/bitwise_test.go
Comment (toni-calvin) [reply pkg/runners/builtin_runner/bitwise_test.go]: True, done, here and in the rest of the builtin code

**Comment by toni-calvin** [comment]
> No description provided.

This PR aims to implement the Bitwise Builtin. 
Changes: 
- Added methods ADD, OR, XOR for Felts and its testing 
- Better error handling: When an error is encountered, is encapsulates with the fields necessary to give extra information and is wrapped with class error. 
- Deserialise program identifiers field. Until now, `value` was expected a number, but is a felt, and in the json is provided as bigInt. Changes were made to do this conversion.

File: pkg/builtins/bitwise.go
Comment (toni-calvin) [inline pkg/builtins/bitwise.go:21]: #144

File: pkg/builtins/bitwise.go
Comment (toni-calvin) [inline pkg/builtins/bitwise.go:1]: #121

File: pkg/builtins/bitwise.go
Comment (toni-calvin) [inline pkg/builtins/bitwise.go]: Error handling example

File: README.md
Comment (gabrielbosio) [reply README.md:2498]: Agree

File: pkg/builtins/bitwise_test.go
Comment (gabrielbosio) [inline pkg/builtins/bitwise_test.go]: I think this line can be removed, because `Memory.Get` returns 0 by default

File: pkg/builtins/bitwise_test.go
Comment (gabrielbosio) [inline pkg/builtins/bitwise_test.go]: I think this line can be removed, because `Memory.Get` returns 0 by default

**Review by gabrielbosio — APPROVED** [review]
Approved with a few comments

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]:
```suggestion
	return BitwiseError(errors.Errorf("Expected felt %d to be smaller than  2**%d", felt, upperBound))
```
This looks clearer to the user imo. Also if the value is constant we can hardcode it in the error message

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]: Here index will never be bigger than addr.Offset so we can ignore the error (with a comment)

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]:
```suggestion
		return nil, nil
```
Correct behaviour

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]:
```suggestion
	y_addr := x_addr.AddUint(1)
	num_y, err := segments.Get(y_addr)
	if err != nil {
		return nil, nil
	}
```
Correct behaviour

File: pkg/builtins/bitwise_test.go
Comment (fmoletta) [reply pkg/builtins/bitwise_test.go]: It doesn't return 0 by default, it returns a zero value when it fails, if we don't insert the 0, Memory.Get will still fail (aka return (0, false))

File: pkg/builtins/bitwise_test.go
Comment (fmoletta) [reply pkg/builtins/bitwise_test.go]: Memory.Get would still fail

File: pkg/parser/parser.go
Comment (fmoletta) [inline pkg/parser/parser.go:51]: Can we use `Felt` here?

File: pkg/runners/cairo_runner.go
Comment (fmoletta) [inline pkg/runners/cairo_runner.go]: why

File: pkg/vm/program.go
Comment (fmoletta) [inline pkg/vm/program.go:47]: Could we use the same Identifier struct for the parser and the program?

File: pkg/builtins/bitwise.go
Comment (toni-calvin) [reply pkg/builtins/bitwise.go]: Why error should not be propagated?

File: pkg/builtins/bitwise.go
Comment (toni-calvin) [reply pkg/builtins/bitwise.go]: 100% agree, I though of doing a more abstract composition if there were other callers, but I agree is clearer

File: pkg/builtins/bitwise_test.go
Comment (toni-calvin) [reply pkg/builtins/bitwise_test.go]: In fact it fails, we are returning this: 

`
value, ok := m.data[addr]
if !ok {
 return nil, errors.New("Memory Get: Value not found")
}
`

But in the tests that position was not being loaded from memory so it did not fail, I add again the values

File: pkg/vm/program.go
Comment (toni-calvin) [reply pkg/vm/program.go:47]: We use `err = json.Unmarshal(byteValue, &cJson)` to unmarshall json into parser struct, but can't define a type felt there. It is possible but we would have to iterate through field to unmarshall them. This way we can point to reference while the structure is the same or add conversion functions when needed.

File: README.md
Comment (fmoletta) [inline README.md:2459]:
```suggestion
func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, memory *memory.Memory) (*memory.MaybeRelocatable, error) {
```

File: README.md
Comment (fmoletta) [inline README.md:2480]:
```suggestion
	num_x, err := segments.Get(x_addr)
	if err != nil {

		return nil, nil
	}

	num_y, err := segments.Get(y_addr)
	if err != nil {

		return nil, nil
	}
```
It shouldn't fail here

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]: We can use `memory.GetFelt` to avoid having to use memory.Get + num.GetFelt

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]:
```suggestion
func (b *BitwiseBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, memory *memory.Memory) (*memory.MaybeRelocatable, error) {
```
Wrong name, `mem` could also be used if it conflicts with the package name

File: pkg/builtins/bitwise.go
Comment (fmoletta) [inline pkg/builtins/bitwise.go]: Remove

File: README.md
Comment (fmoletta) [inline README.md]: Please update this

File: README.md
Comment (fmoletta) [inline README.md]: Remove

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**
